tools/skbstat.py

```diff
@@ -0,0 +1,1015 @@
+	    *((unsigned __int128 *)(filter->addr6[1]));
+	unsigned __int128 saddr_mask =
+	    *((unsigned __int128 *)(filter->mask6[0])), daddr_mask =
+	    *((unsigned __int128 *)(filter->mask6[1]));
+
+	return (saddr_expect ? (saddr & saddr_mask) == saddr_expect : 1)
+	    && (daddr_expect ? (daddr & daddr_mask) == daddr_expect : 1);
+}
+
+static inline int ip6_addr_match1(unsigned __int128 saddr,
+				  unsigned __int128 daddr,
+				  struct filter *filter)
+{
+	unsigned __int128 addr_expect =
+	    *((unsigned __int128 *)(filter->addr6[0]));
+	unsigned __int128 addr_mask =
+	    *((unsigned __int128 *)(filter->mask6[0]));
+
+	return ((saddr & addr_mask) == addr_expect)
+	    || ((daddr & addr_mask) == addr_expect);
+}
+
+static inline int ip6_addr_match(unsigned __int128 saddr,
+				 unsigned __int128 daddr, struct filter *filter)
+{
+	return (filter->flags & SKB_ADDR_MATCH_1) ? ip6_addr_match1(saddr,
+								    daddr,
+								    filter) :
+	    ip6_addr_match2(saddr, daddr, filter);
+}
+
+static inline int ip_addr_match2(__be32 saddr, __be32 daddr,
+				 struct filter *filter)
+{
+	return (filter->addr[0] ? (saddr & filter->mask[0]) ==
+		filter->addr[0] : 1)
+	    && (filter->addr[1] ? (daddr & filter->mask[1]) ==
+		filter->addr[1] : 1);
+}
+
+static inline int ip_addr_match1(__be32 saddr, __be32 daddr,
+				 struct filter *filter)
+{
+	return ((saddr & filter->mask[0]) == filter->addr[0])
+	    || ((daddr & filter->mask[0]) == filter->addr[0]);
+}
+
+static inline int ip_addr_match(__be32 saddr, __be32 daddr,
+				struct filter *filter)
+{
+	return (filter->flags & SKB_ADDR_MATCH_1) ? ip_addr_match1(saddr, daddr,
+								   filter) :
+	    ip_addr_match2(saddr, daddr, filter);
+}
+
+static inline int record_skb(struct pt_regs *ctx, union filter_args *args)
+{
+	u64 *count = NULL, zero = 0, ip = PT_REGS_IP(ctx);
+
+	if (args->input) {
+		count = input.lookup_or_try_init(&ip, &zero);
+	} else if (args->output) {
+		count = output.lookup_or_try_init(&ip, &zero);
+	} else if (args->forward) {
+		count = forward.lookup_or_try_init(&ip, &zero);
+	}
+
+	if (count) {
+		*count += 1;
+	}
+
+	return 0;
+}
+
+static inline int record_sk(struct pt_regs *ctx, union filter_args *args)
+{
+	return record_skb(ctx, args);
+}
+
+static inline int record_skb_backtrace(struct pt_regs *ctx,
+				       union filter_args *args)
+{
+	u32 stack_id = stack_traces.get_stackid(ctx, 0);
+	u64 *count, zero = 0;
+
+	if (args->drop) {
+		count = drop_stacks.lookup_or_try_init(&stack_id, &zero);
+	} else {
+		count = consume_stacks.lookup_or_try_init(&stack_id, &zero);
+	}
+
+	if (count) {
+		*count += 1;
+	}
+
+	return 0;
+}
+
+static inline int skb_l4_match(u8 proto, union l4hdr *hdr,
+			       struct filter *filter)
+{
+	switch (proto) {
+	case IPPROTO_TCP:
+	case IPPROTO_UDP:
+		{
+			u16 dst = hdr->tcpudphdr.dst, src = hdr->tcpudphdr.src;
+
+			return port_match(src, dst, filter);
+		}
+	default:
+		break;
+	}
+	return 1;
```

Comment:
So any other protocols will consider as match?

Reply:
`skb_l4_match` will check the detail of L4 protocol, such as TCP port, ICMP id. Protocols as ICMP will just return 1 and skip the L4 check currently. Protocol number will be checked in `proto_match` function.